### ISSUE TITLE
🚀 Graceful RPCAgent termination in multi-driver scenario

### ISSUE THREAD

**Issue by xush6528** [pytorch/pytorch#27647]
# Problem

* Currently, `sync()`, `join()` interface is implemented in and only for `ProcessGroupAgent`.
* Other `RPCAgent`s also need it, but it shouldn’t implement it again.
* What stops the current solution from being reused is that `ProcessGroupAgent` uses collective communication to achieve it, while by default, other `RPCAgent`s don't have this facility.

# Proposal

* Problem 1
    * Statement
        * Currently, we ask users to call `init_process_group` before calling `init_model_parallel`. One user error could be that the user passes 2 different init_methods, resulting 2 different node groups. One for c10d data parallel default process group and the other for model parallel node group.
    * Solution
        * We create a model parallel default process group inside of `init_model_parallel` for users. Notice, the default process group here is a model-parallel scope default. In order to make naming scopes of data parallelism and model parallelism distinct. We decide to do the following refactoring.
            * Move `torch.distributed.*`, like `torch.distributed.barrier`, to `torch.distributed.data_parallel.*`.
            * Keep `torch.distributed.rpc.*` in case data parallel want to use RPC in the future.
            * Create `torch.distributed.model_parallel.*`. Inside it, it could import `torch.distributed.rpc` to use.
        * The `init_model_parallel` API initializes a model parallel default process group for users. Users can call `torch.distributed.model_parallel.barrier` later.
* Problem 2
    * Statement
        * Is `ProcessGroupAgent::sync` still needed? It essentially called `barrier()` twice. Shouldn’t `barrier` + `drain` enough to gracefully shutdown symmetric workers (i.e. workers are not driven by a master)?
    * Solution
        * Remove `sync`, `join` interface in `RpcAgent`. For user cases where there is no master in charge, make every worker call `barrier` + `drain` at the end, before shutting down local `RPCAgent`.

**Comment by mrshenli** [comment]
> Move torch.distributed.*, like torch.distributed.barrier, to torch.distributed.data_parallel.*.

I am not sure if we would like to use the name `data_parallel` for this. Those are c10d APIs, and data parallel and model parallel are layers on top of that. Upper layers could use `c10d::allreduce` to build DDP, or model averaging, or use `c10d::send/recv` to build model parallel (like we did in ProcessGroupAgent). I looks to me `torch.distributed.*` is an appropriate place for those APIs. If we have to move them around, maybe `torch.distributed.c10d.*`? 

> Keep torch.distributed.rpc.* in case data parallel want to use RPC in the future.

What are these APIs? Are you referring to sync and join?

> Create torch.distributed.model_parallel.*. Inside it, it could import torch.distributed.rpc to use.

Similar to the above comments for `data_parallel`, I feel `model_parallel` might not be a package name for those APIs. And as we have discussed in #27290, we might want to rename `init_model_parallel` to `init_worker` or just `init`.

> Currently, we ask users to call init_process_group before calling init_model_parallel. One user error could be that the user passes 2 different init_methods, resulting 2 different node groups. One for c10d data parallel default process group and the other for model parallel node group.

I agree. We should decouple `init_process_group` and `init_model_parallel`. Would it be sufficient if `init_model_parallel` creates a new set of process group objects internally?

**Comment by xush6528** [comment]
> Move torch.distributed., like torch.distributed.barrier, to torch.distributed.data_parallel..

No problem. I can leave `torch.distributed.*` as it is.

> Keep torch.distributed.rpc.* in case data parallel want to use RPC in the future.

remote, rpc_sync/async, sync, join.

> Create torch.distributed.model_parallel.*. Inside it, it could import torch.distributed.rpc to use.

Currently, `init_model_parallel` has 3 steps. 
  1) create rpc agent. 
  2) initialize rref context.
  3) initialize dist autograd context.

I think we should add `rpc.init_worker` is essnetially `rpc._init_rpc`, we should rename it to `rpc._init` to match `distautograd._init`. Also has `rref._init()`.

Agree?

> Currently, we ask users to call init_process_group before calling init_model_parallel. One user error could be that the user passes 2 different init_methods, resulting 2 different node groups. One for c10d data parallel default process group and the other for model parallel node group.

Agree. I plan to create process group internally, inside of `init_model_parallel `. 
I'm going to have a PR out soon.

@mrshenli

**Comment by mrshenli** [comment]
> I think we should add rpc.init_worker is essnetially rpc._init_rpc, we should rename it to rpc._init to match distautograd._init. Also has rref._init().

Yep, this sounds good to me. 

> Is ProcessGroupAgent::sync still needed? It essentially called barrier() twice. 

I think we can remove `sync` and then move `join` out of `RpcAgent` to be a utility function. 

> Shouldn’t barrier + drain enough to gracefully shutdown symmetric workers (i.e. workers are not driven by a master)?



(Assuming that `drain()` blocks until no more tasks in thread pool, and until all local futures are done) That requires applications to wait on every futures created in nested rpc/remote, even if they don't care about some responses, otherwise it could hit timeout or connection closed by peer when processing nested RPCs. As we don't have an API for users to implement async UDFs, the implication for the above is that all rpc/remote calls will be blocking on the server side. We need to be very clear about this in the doc, and mention that if applications would like to make async rpc call and not waiting for the response, they have to call the `join` utility function at the end. 

This thing becomes a little trickier with RRefs. Consider the following code:

```python
# on worker 0
def func() -> RRef:
    rref = rpc.remote("worker0", func_with_rpc, args=(1, 2))
    # A future will be created internally, but users don't have access to it. 
    # Does this mean we need to let RRef wrap the future and expose a wait() 
    # API on RRef? Above doesn't look like a friendly API. Consider `Model.__init__()` 
    # creates multiple RRefs that points to different subnets, and we replace 
    # `func_with_rpc` with `Model` here.  That `Model.__init__()` method definitely 
    # should not wait block until all RRefs are ready. Then it means that the caller of
    # the model constructor has  to go in and retrieve all RRefs and call wait on all of 
    # them. 
    return rref

x = rpc.remote("worker1", func)

barrier()
# it is possible that the remote message has not been sent yet
drain()
# all messages are sent, and local futures are done. If we don't block on RRef, the message
# from worker1 to worker0 could arrive after drain, which will trigger another message been
# sent, and we won't do drain again. But blocking on RRef will become messy with complex 
# user data types. 

# on worker 1
barrier()
```

BTW, which barrier are you going to use here? Is it `c10d::barrier`, or is it implemented by `RpcAgent` implementations?

**Comment by rohan-varma** [comment]
cc @jiayisuse Is this issue still relevant? Feel free to close if not.
